{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "sufficient-fashion",
   "metadata": {},
   "source": [
    "The goal of this analysis is to try and refine the 3' end calling of TALON by using the GENCODE strategy of longest TES per transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "equal-electric",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "modified-struggle",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_longest_ends(annot, how='tes', novelty='novel'):\n",
    "    df = pd.read_csv(annot, sep='\\t')\n",
    "    \n",
    "    if novelty == 'novel':\n",
    "        df = df.loc[df.transcript_novelty != 'Known']\n",
    "    \n",
    "    fwd = df.loc[df.strand == '+']\n",
    "    rev = df.loc[df.strand == '-']\n",
    "    \n",
    "    # furthest downstream for tes\n",
    "    # if + strand, max coord of read end\n",
    "    # if - strand, min coord of read end\n",
    "    if how == 'tes':\n",
    "        fwd = fwd[['transcript_ID', 'read_end']]\n",
    "        fwd = fwd.groupby('transcript_ID').max().reset_index()\n",
    "        rev = rev[['transcript_ID', 'read_end']]\n",
    "        rev = rev.groupby('transcript_ID').min().reset_index()\n",
    "        \n",
    "        \n",
    "    # furthest upstream for tss:\n",
    "    # if + strand, min coord of read start\n",
    "    # if - strand, max coord of read start\n",
    "    elif how == 'tss':\n",
    "        fwd = fwd[['transcript_ID', 'read_start']]\n",
    "        fwd = fwd.groupby('transcript_ID').min().reset_index()\n",
    "        rev = rev[['transcript_ID', 'read_start']]\n",
    "        rev = rev.groupby('transcript_ID').max().reset_index()\n",
    "    \n",
    "    # concat fwd and rev\n",
    "    df = pd.concat([fwd, rev])\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fabulous-tunnel",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace tss or tes coords with those in a different df\n",
    "# gtf: gtf file location \n",
    "# ends: df with transcript_ID, end coordinate\n",
    "# how: 'tss' or 'tes'\n",
    "# opref: output file prefix\n",
    "def replace_gtf_end_coords(gtf, ends, opref, how='tes', test=False):\n",
    "    \n",
    "    # read preexisting GTF and \n",
    "    gtf_df = pd.read_csv(gtf, sep='\\t', header=None, \\\n",
    "                names=['chr', 'source', 'entry_type', \\\n",
    "                       'start', 'stop', 'score', 'strand',\\\n",
    "                       'frame', 'fields'], comment='#')\n",
    "\n",
    "    # get relevant values from fields\n",
    "    gtf_df['transcript_id'] = np.nan\n",
    "    gtf_df.loc[gtf_df.entry_type!='gene', 'transcript_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].fields.str.split(pat='talon_transcript \"', n=1, expand=True)[1]\n",
    "    gtf_df.loc[gtf_df.entry_type!='gene', 'transcript_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].transcript_id.str.split(pat='\"', n=1, expand=True)[0]\n",
    "    \n",
    "\n",
    "    if how == 'tes':\n",
    "        ends.columns = ['transcript_id', 'tes']\n",
    "    elif how == 'tss':\n",
    "        ends.columns = ['transcript_id', 'tss']\n",
    "        \n",
    "    # merge gtf_df with end information \n",
    "#     ends.transcript_id = ends.transcript_id.astype('str')\n",
    "    df = gtf_df.loc[gtf_df.transcript_id.notnull()]\n",
    "    ends.transcript_id = ends.transcript_id.astype('str')\n",
    "    df.transcript_id = df.transcript_id.astype('str')\n",
    "    gtf_df = gtf_df.merge(ends, how='left', on='transcript_id')\n",
    "    \n",
    "    if test:\n",
    "        tids = ['142372', '142634']\n",
    "        print('142372' in df.transcript_id)\n",
    "        df = df.loc[df.transcript_id.isin(tids)]\n",
    "        print(ends.loc[ends.transcript_id.isin(tids)])\n",
    "        print()\n",
    "        print('Before editing')\n",
    "        print(gtf_df[['transcript_id', 'entry_type', 'strand', 'start', 'stop', how]])\n",
    "    \n",
    "    # swap out read starts or ends for the longest ones\n",
    "    # fwd: swap out transcript \"stop\" and last exon \"stop\" \n",
    "    # rev: swap out transcript \"start\" and last exon \"start\"\n",
    "    tids = df.transcript_id.unique()\n",
    "    for t, tid in enumerate(tids):\n",
    "        if t % 1000 == 0:\n",
    "            print('Processing transcript {} of {}'.format(t, len(tids)))\n",
    "        if how == 'tes':\n",
    "            # transcript entry for fwd\n",
    "            ind = gtf_df.loc[(gtf_df.strand=='+')&(gtf_df.transcript_id==tid)].index.tolist()\n",
    "            if test:\n",
    "                print('fwd ind')\n",
    "                print(ind)\n",
    "            if ind: \n",
    "                if test: \n",
    "                    print('fixing...')\n",
    "                i = ind[0]\n",
    "                gtf_df.loc[i, 'stop'] = gtf_df.loc[i, 'tes']\n",
    "                # stop of last exon for fwd\n",
    "                i = ind[-1]\n",
    "                gtf_df.loc[i, 'stop'] = gtf_df.loc[i, 'tes']\n",
    "\n",
    "            # transcript for rev\n",
    "            ind = gtf_df.loc[(gtf_df.strand=='-')&(gtf_df.transcript_id==tid)].index.tolist()\n",
    "            if test: \n",
    "                print('rev ind')\n",
    "                print(ind)\n",
    "            if ind: \n",
    "                if test:\n",
    "                    print('fixing...')\n",
    "                i = ind[0]\n",
    "                gtf_df.loc[i, 'start'] = gtf_df.loc[i, 'tes']\n",
    "                # start of last exon for rev\n",
    "                i = ind[-1]\n",
    "                gtf_df.loc[i, 'start'] = gtf_df.loc[i, 'tes']\n",
    "\n",
    "            # for exon can probably just take [1] or [-1] of temp to access\n",
    "            # first or last exon respectively\n",
    "\n",
    "        # fwd: swap out transcript \"start\" and first exon \"start\" \n",
    "        # rev: swap out transcript \"stop\" and first exon \"stop\"                 \n",
    "#         elif how == 'tss':\n",
    "#             # transcript entry for fwd\n",
    "#             ind = df.loc[(df.strand=='+')&(df.transcript_id==tid)].index.tolist()\n",
    "#             if ind: \n",
    "#                 df.loc[ind, 'start'] = df.loc[ind, 'tss']\n",
    "#                 # start of first exon for fwd\n",
    "#                 ind = ind[1]\n",
    "#                 df.loc[ind].start = df.loc[ind].tss\n",
    "#             # transcript entry for rev\n",
    "#             ind = df.loc[(df.strand=='-')&(df.transcript_id==tid)].index.tolist()\n",
    "#             if ind:\n",
    "#                 df.loc[ind].stop = df.loc[ind].tss\n",
    "#                 # stop of first exon for rev\n",
    "#                 ind = ind[1]\n",
    "#                 df.loc[ind].stop = df.loc[ind].tss\n",
    "\n",
    "    if test:\n",
    "        print()\n",
    "        print('After editing')\n",
    "        print(gtf_df[['transcript_id', 'entry_type', 'strand', 'start', 'stop', how]])\n",
    "      \n",
    "    cols=['chr', 'source', 'entry_type', \\\n",
    "          'start', 'stop', 'score', 'strand',\\\n",
    "           'frame', 'fields']\n",
    "    gtf_df = gtf_df[cols]\n",
    "    gtf_df['start'] = gtf_df['start'].astype('int')\n",
    "    gtf_df['stop'] = gtf_df['stop'].astype('int')\n",
    "    if test:\n",
    "        fname = '{}_revised_tes_test.gtf'.format(opref)\n",
    "    else:\n",
    "        fname = '{}_revised_tes.gtf'.format(opref)\n",
    "    gtf_df.to_csv(fname, sep='\\t', header=None, index=False, quoting=csv.QUOTE_NONE)\n",
    "    return gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "introductory-tokyo",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'test.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "desperate-survival",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "    transcript_id      tes\n",
      "229        142372  4899923\n",
      "304        142634  3344252\n",
      "\n",
      "Before editing\n",
      "   transcript_id  entry_type strand    start     stop        tes\n",
      "0            NaN        gene      +  4857814  4897905        NaN\n",
      "1         142372  transcript      +  4857814  4897905  4899923.0\n",
      "2         142372        exon      +  4857814  4857976  4899923.0\n",
      "3         142372        exon      +  4867470  4867532  4899923.0\n",
      "4         142372        exon      +  4878027  4878132  4899923.0\n",
      "5         142372        exon      +  4886744  4886831  4899923.0\n",
      "6         142372        exon      +  4889460  4889602  4899923.0\n",
      "7         142372        exon      +  4890740  4890796  4899923.0\n",
      "8         142372        exon      +  4891915  4892069  4899923.0\n",
      "9         142372        exon      +  4893417  4893563  4899923.0\n",
      "10        142372        exon      +  4894934  4895005  4899923.0\n",
      "11        142372        exon      +  4896356  4897905  4899923.0\n",
      "12           NaN        gene      -  3344256  3409189        NaN\n",
      "13        142634  transcript      -  3344256  3371376  3344252.0\n",
      "14        142634        exon      -  3370969  3371376  3344252.0\n",
      "15        142634        exon      -  3360451  3360586  3344252.0\n",
      "16        142634        exon      -  3358995  3359104  3344252.0\n",
      "17        142634        exon      -  3358110  3358298  3344252.0\n",
      "18        142634        exon      -  3355811  3355916  3344252.0\n",
      "19        142634        exon      -  3355273  3355469  3344252.0\n",
      "20        142634        exon      -  3352291  3353463  3344252.0\n",
      "21        142634        exon      -  3350642  3350773  3344252.0\n",
      "22        142634        exon      -  3350392  3350448  3344252.0\n",
      "23        142634        exon      -  3348594  3348659  3344252.0\n",
      "24        142634        exon      -  3347574  3347748  3344252.0\n",
      "25        142634        exon      -  3347052  3347107  3344252.0\n",
      "26        142634        exon      -  3346235  3346392  3344252.0\n",
      "27        142634        exon      -  3344256  3345884  3344252.0\n",
      "28           NaN        gene      -  3206523  3215632        NaN\n",
      "29             4  transcript      -  3206523  3215632        NaN\n",
      "30             4        exon      -  3213439  3215632        NaN\n",
      "31             4        exon      -  3206523  3207317        NaN\n",
      "Processing transcript 0 of 2\n",
      "fwd ind\n",
      "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]\n",
      "fixing...\n",
      "rev ind\n",
      "[]\n",
      "fwd ind\n",
      "[]\n",
      "rev ind\n",
      "[13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27]\n",
      "fixing...\n",
      "\n",
      "After editing\n",
      "   transcript_id  entry_type strand      start       stop        tes\n",
      "0            NaN        gene      +  4857814.0  4897905.0        NaN\n",
      "1         142372  transcript      +  4857814.0  4899923.0  4899923.0\n",
      "2         142372        exon      +  4857814.0  4857976.0  4899923.0\n",
      "3         142372        exon      +  4867470.0  4867532.0  4899923.0\n",
      "4         142372        exon      +  4878027.0  4878132.0  4899923.0\n",
      "5         142372        exon      +  4886744.0  4886831.0  4899923.0\n",
      "6         142372        exon      +  4889460.0  4889602.0  4899923.0\n",
      "7         142372        exon      +  4890740.0  4890796.0  4899923.0\n",
      "8         142372        exon      +  4891915.0  4892069.0  4899923.0\n",
      "9         142372        exon      +  4893417.0  4893563.0  4899923.0\n",
      "10        142372        exon      +  4894934.0  4895005.0  4899923.0\n",
      "11        142372        exon      +  4896356.0  4899923.0  4899923.0\n",
      "12           NaN        gene      -  3344256.0  3409189.0        NaN\n",
      "13        142634  transcript      -  3344252.0  3371376.0  3344252.0\n",
      "14        142634        exon      -  3370969.0  3371376.0  3344252.0\n",
      "15        142634        exon      -  3360451.0  3360586.0  3344252.0\n",
      "16        142634        exon      -  3358995.0  3359104.0  3344252.0\n",
      "17        142634        exon      -  3358110.0  3358298.0  3344252.0\n",
      "18        142634        exon      -  3355811.0  3355916.0  3344252.0\n",
      "19        142634        exon      -  3355273.0  3355469.0  3344252.0\n",
      "20        142634        exon      -  3352291.0  3353463.0  3344252.0\n",
      "21        142634        exon      -  3350642.0  3350773.0  3344252.0\n",
      "22        142634        exon      -  3350392.0  3350448.0  3344252.0\n",
      "23        142634        exon      -  3348594.0  3348659.0  3344252.0\n",
      "24        142634        exon      -  3347574.0  3347748.0  3344252.0\n",
      "25        142634        exon      -  3347052.0  3347107.0  3344252.0\n",
      "26        142634        exon      -  3346235.0  3346392.0  3344252.0\n",
      "27        142634        exon      -  3344252.0  3345884.0  3344252.0\n",
      "28           NaN        gene      -  3206523.0  3215632.0        NaN\n",
      "29             4  transcript      -  3206523.0  3215632.0        NaN\n",
      "30             4        exon      -  3213439.0  3215632.0        NaN\n",
      "31             4        exon      -  3206523.0  3207317.0        NaN\n"
     ]
    }
   ],
   "source": [
    "df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes', test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "conceptual-precipitation",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "     transcript_id      tes\n",
      "229         142372  4899923\n",
      "304         142634  3344252\n",
      "\n",
      "Before editing\n",
      "        transcript_id  entry_type strand    start     stop        tes\n",
      "92             142372  transcript      +  4857814  4897905  4899923.0\n",
      "93             142372        exon      +  4857814  4857976  4899923.0\n",
      "94             142372        exon      +  4867470  4867532  4899923.0\n",
      "95             142372        exon      +  4878027  4878132  4899923.0\n",
      "96             142372        exon      +  4886744  4886831  4899923.0\n",
      "97             142372        exon      +  4889460  4889602  4899923.0\n",
      "98             142372        exon      +  4890740  4890796  4899923.0\n",
      "99             142372        exon      +  4891915  4892069  4899923.0\n",
      "100            142372        exon      +  4893417  4893563  4899923.0\n",
      "101            142372        exon      +  4894934  4895005  4899923.0\n",
      "102            142372        exon      +  4896356  4897905  4899923.0\n",
      "228081         142634  transcript      -  3344256  3371376  3344252.0\n",
      "228082         142634        exon      -  3370969  3371376  3344252.0\n",
      "228083         142634        exon      -  3360451  3360586  3344252.0\n",
      "228084         142634        exon      -  3358995  3359104  3344252.0\n",
      "228085         142634        exon      -  3358110  3358298  3344252.0\n",
      "228086         142634        exon      -  3355811  3355916  3344252.0\n",
      "228087         142634        exon      -  3355273  3355469  3344252.0\n",
      "228088         142634        exon      -  3352291  3353463  3344252.0\n",
      "228089         142634        exon      -  3350642  3350773  3344252.0\n",
      "228090         142634        exon      -  3350392  3350448  3344252.0\n",
      "228091         142634        exon      -  3348594  3348659  3344252.0\n",
      "228092         142634        exon      -  3347574  3347748  3344252.0\n",
      "228093         142634        exon      -  3347052  3347107  3344252.0\n",
      "228094         142634        exon      -  3346235  3346392  3344252.0\n",
      "228095         142634        exon      -  3344256  3345884  3344252.0\n",
      "Processing transcript 0 of 2\n",
      "fwd ind\n",
      "[92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102]\n",
      "fixing...\n",
      "rev ind\n",
      "[]\n",
      "fwd ind\n",
      "[]\n",
      "rev ind\n",
      "[228081, 228082, 228083, 228084, 228085, 228086, 228087, 228088, 228089, 228090, 228091, 228092, 228093, 228094, 228095]\n",
      "%%fileixing...\n",
      "\n",
      "After editing\n",
      "        transcript_id  entry_type strand      start       stop        tes\n",
      "92             142372  transcript      +  4857814.0  4899923.0  4899923.0\n",
      "93             142372        exon      +  4857814.0  4857976.0  4899923.0\n",
      "94             142372        exon      +  4867470.0  4867532.0  4899923.0\n",
      "95             142372        exon      +  4878027.0  4878132.0  4899923.0\n",
      "96             142372        exon      +  4886744.0  4886831.0  4899923.0\n",
      "97             142372        exon      +  4889460.0  4889602.0  4899923.0\n",
      "98             142372        exon      +  4890740.0  4890796.0  4899923.0\n",
      "99             142372        exon      +  4891915.0  4892069.0  4899923.0\n",
      "100            142372        exon      +  4893417.0  4893563.0  4899923.0\n",
      "101            142372        exon      +  4894934.0  4895005.0  4899923.0\n",
      "102            142372        exon      +  4896356.0  4899923.0  4899923.0\n",
      "228081         142634  transcript      -  3344252.0  3371376.0  3344252.0\n",
      "228082         142634        exon      -  3370969.0  3371376.0  3344252.0\n",
      "228083         142634        exon      -  3360451.0  3360586.0  3344252.0\n",
      "228084         142634        exon      -  3358995.0  3359104.0  3344252.0\n",
      "228085         142634        exon      -  3358110.0  3358298.0  3344252.0\n",
      "228086         142634        exon      -  3355811.0  3355916.0  3344252.0\n",
      "228087         142634        exon      -  3355273.0  3355469.0  3344252.0\n",
      "228088         142634        exon      -  3352291.0  3353463.0  3344252.0\n",
      "228089         142634        exon      -  3350642.0  3350773.0  3344252.0\n",
      "228090         142634        exon      -  3350392.0  3350448.0  3344252.0\n",
      "228091         142634        exon      -  3348594.0  3348659.0  3344252.0\n",
      "228092         142634        exon      -  3347574.0  3347748.0  3344252.0\n",
      "228093         142634        exon      -  3347052.0  3347107.0  3344252.0\n",
      "228094         142634        exon      -  3346235.0  3346392.0  3344252.0\n",
      "228095         142634        exon      -  3344252.0  3345884.0  3344252.0\n"
     ]
    }
   ],
   "source": [
    "# annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "# gtf = 'Brain_talon.gtf'\n",
    "\n",
    "# tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "# df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes', test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "simple-least",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing transcript 0 of 37457\n",
      "Processing transcript 1000 of 37457\n",
      "Processing transcript 2000 of 37457\n",
      "Processing transcript 3000 of 37457\n",
      "Processing transcript 4000 of 37457\n",
      "Processing transcript 5000 of 37457\n",
      "Processing transcript 6000 of 37457\n",
      "Processing transcript 7000 of 37457\n",
      "Processing transcript 8000 of 37457\n",
      "Processing transcript 9000 of 37457\n",
      "Processing transcript 10000 of 37457\n",
      "Processing transcript 11000 of 37457\n",
      "Processing transcript 12000 of 37457\n",
      "Processing transcript 13000 of 37457\n",
      "Processing transcript 14000 of 37457\n",
      "Processing transcript 15000 of 37457\n",
      "Processing transcript 16000 of 37457\n",
      "Processing transcript 17000 of 37457\n",
      "Processing transcript 18000 of 37457\n",
      "Processing transcript 19000 of 37457\n",
      "Processing transcript 20000 of 37457\n",
      "Processing transcript 21000 of 37457\n",
      "Processing transcript 22000 of 37457\n",
      "Processing transcript 23000 of 37457\n",
      "Processing transcript 24000 of 37457\n",
      "Processing transcript 25000 of 37457\n",
      "Processing transcript 26000 of 37457\n",
      "Processing transcript 27000 of 37457\n",
      "Processing transcript 28000 of 37457\n",
      "Processing transcript 29000 of 37457\n",
      "Processing transcript 30000 of 37457\n",
      "Processing transcript 31000 of 37457\n",
      "Processing transcript 32000 of 37457\n",
      "Processing transcript 33000 of 37457\n",
      "Processing transcript 34000 of 37457\n",
      "Processing transcript 35000 of 37457\n",
      "Processing transcript 36000 of 37457\n",
      "Processing transcript 37000 of 37457\n"
     ]
    }
   ],
   "source": [
    "# cortex\n",
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'mouse_brain_cortex_talon.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "df = replace_gtf_end_coords(gtf, tes, 'mouse_brain_cortex', how='tes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "miniature-jacket",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing transcript 0 of 37034\n",
      "Processing transcript 1000 of 37034\n",
      "Processing transcript 2000 of 37034\n",
      "Processing transcript 3000 of 37034\n",
      "Processing transcript 4000 of 37034\n",
      "Processing transcript 5000 of 37034\n",
      "Processing transcript 6000 of 37034\n",
      "Processing transcript 7000 of 37034\n",
      "Processing transcript 8000 of 37034\n",
      "Processing transcript 9000 of 37034\n",
      "Processing transcript 10000 of 37034\n",
      "Processing transcript 11000 of 37034\n",
      "Processing transcript 12000 of 37034\n",
      "Processing transcript 13000 of 37034\n",
      "Processing transcript 14000 of 37034\n",
      "Processing transcript 15000 of 37034\n",
      "Processing transcript 16000 of 37034\n",
      "Processing transcript 17000 of 37034\n",
      "Processing transcript 18000 of 37034\n",
      "Processing transcript 19000 of 37034\n",
      "Processing transcript 20000 of 37034\n",
      "Processing transcript 21000 of 37034\n",
      "Processing transcript 22000 of 37034\n",
      "Processing transcript 23000 of 37034\n",
      "Processing transcript 24000 of 37034\n",
      "Processing transcript 25000 of 37034\n",
      "Processing transcript 26000 of 37034\n",
      "Processing transcript 27000 of 37034\n",
      "Processing transcript 28000 of 37034\n",
      "Processing transcript 29000 of 37034\n",
      "Processing transcript 30000 of 37034\n",
      "Processing transcript 31000 of 37034\n",
      "Processing transcript 32000 of 37034\n",
      "Processing transcript 33000 of 37034\n",
      "Processing transcript 34000 of 37034\n",
      "Processing transcript 35000 of 37034\n",
      "Processing transcript 36000 of 37034\n",
      "Processing transcript 37000 of 37034\n"
     ]
    }
   ],
   "source": [
    "# hippocampus\n",
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'mouse_brain_hippocampus_talon.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "df = replace_gtf_end_coords(gtf, tes, 'mouse_brain_hippocampus', how='tes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "senior-representative",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix stupid numbors\n",
    "gtf = 'mouse_brain_cortex_revised_tes.gtf'\n",
    "gtf_df = pd.read_csv(gtf, sep='\\t', header=None, \\\n",
    "            names=['chr', 'source', 'entry_type', \\\n",
    "                   'start', 'stop', 'score', 'strand',\\\n",
    "                   'frame', 'fields'])\n",
    "gtf_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
