{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "national-mandate",
   "metadata": {},
   "source": [
    "The goal of this analysis is to try and refine the 3' end calling of TALON by using the GENCODE strategy of longest TES per transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "whole-centre",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "aggressive-grant",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_longest_ends(annot, how='tes', novelty='novel'):\n",
    "    df = pd.read_csv(annot, sep='\\t')\n",
    "    \n",
    "    if novelty == 'novel':\n",
    "        df = df.loc[df.transcript_novelty != 'Known']\n",
    "    \n",
    "    fwd = df.loc[df.strand == '+']\n",
    "    rev = df.loc[df.strand == '-']\n",
    "    \n",
    "    # furthest downstream for tes\n",
    "    # if + strand, max coord of read end\n",
    "    # if - strand, min coord of read end\n",
    "    if how == 'tes':\n",
    "        fwd = fwd[['transcript_ID', 'read_end']]\n",
    "        fwd = fwd.groupby('transcript_ID').max().reset_index()\n",
    "        rev = rev[['transcript_ID', 'read_end']]\n",
    "        rev = rev.groupby('transcript_ID').min().reset_index()\n",
    "        \n",
    "        \n",
    "    # furthest upstream for tss:\n",
    "    # if + strand, min coord of read start\n",
    "    # if - strand, max coord of read start\n",
    "    elif how == 'tss':\n",
    "        fwd = fwd[['transcript_ID', 'read_start']]\n",
    "        fwd = fwd.groupby('transcript_ID').min().reset_index()\n",
    "        rev = rev[['transcript_ID', 'read_start']]\n",
    "        rev = rev.groupby('transcript_ID').max().reset_index()\n",
    "    \n",
    "    # concat fwd and rev\n",
    "    df = pd.concat([fwd, rev])\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "tutorial-analyst",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace tss or tes coords with those in a different df\n",
    "# gtf: gtf file location \n",
    "# ends: df with transcript_ID, end coordinate\n",
    "# how: 'tss' or 'tes'\n",
    "# opref: output file prefix\n",
    "def replace_gtf_end_coords(gtf, ends, opref, how='tes', test=False):\n",
    "    \n",
    "    # read preexisting GTF and \n",
    "    gtf_df = pd.read_csv(gtf, sep='\\t', header=None, \\\n",
    "                names=['chr', 'source', 'entry_type', \\\n",
    "                       'start', 'stop', 'score', 'strand',\\\n",
    "                       'frame', 'fields'])\n",
    "\n",
    "    # get relevant values from fields\n",
    "    gtf_df['transcript_id'] = np.nan\n",
    "    gtf_df.loc[gtf_df.entry_type!='gene', 'transcript_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].fields.str.split(pat='talon_transcript \"', n=1, expand=True)[1]\n",
    "    gtf_df.loc[gtf_df.entry_type!='gene', 'transcript_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].transcript_id.str.split(pat='\"', n=1, expand=True)[0]\n",
    "\n",
    "    if how == 'tes':\n",
    "        ends.columns = ['transcript_id', 'tes']\n",
    "    elif how == 'tss':\n",
    "        ends.columns = ['transcript_id', 'tss']\n",
    "        \n",
    "    # merge gtf_df with end information \n",
    "#     ends.transcript_id = ends.transcript_id.astype('str')\n",
    "#     gtf_df.transcript_id = gtf_df.transcript_id.astype('str')\n",
    "    df = gtf_df.loc[gtf_df.transcript_id.notnull()]\n",
    "    ends.transcript_id = ends.transcript_id.astype('int')\n",
    "    df.transcript_id = df.transcript_id.astype('int')\n",
    "    df = df.merge(ends, how='left', on='transcript_id')\n",
    "    \n",
    "    if test:\n",
    "        tids = [142372, 142634]\n",
    "        print(142372 in df.transcript_id)\n",
    "        df = df.loc[df.transcript_id.isin(tids)]\n",
    "        print(ends.loc[ends.transcript_id.isin(tids)])\n",
    "        print()\n",
    "        print('Before editing')\n",
    "        print(df[['transcript_id', 'entry_type', 'strand', 'start', 'stop', how]])\n",
    "    \n",
    "    # swap out read starts or ends for the longest ones\n",
    "    # fwd: swap out transcript \"stop\" and last exon \"stop\" \n",
    "    # rev: swap out transcript \"start\" and last exon \"start\"\n",
    "    for tid in df.transcript_id.unique():\n",
    "        if how == 'tes':\n",
    "            # transcript entry for fwd\n",
    "            ind = df.loc[(df.strand=='+')&(df.transcript_id==tid)].index.tolist()\n",
    "            if test:\n",
    "                print('fwd ind')\n",
    "                print(ind)\n",
    "            if ind: \n",
    "                if test: \n",
    "                    print('fixing...')\n",
    "                i = ind[0]\n",
    "                df.loc[i, 'stop'] = df.loc[i, 'tes']\n",
    "                # stop of last exon for fwd\n",
    "                i = ind[-1]\n",
    "                df.loc[i, 'stop'] = df.loc[i, 'tes']\n",
    "\n",
    "            # transcript for rev\n",
    "            ind = df.loc[(df.strand=='-')&(df.transcript_id==tid)].index.tolist()\n",
    "            if test: \n",
    "                print('rev ind')\n",
    "                print(ind)\n",
    "            if ind: \n",
    "                if test:\n",
    "                    print('%%fileixing...')\n",
    "                i = ind[0]\n",
    "                df.loc[i, 'start'] = df.loc[i, 'tes']\n",
    "                # start of last exon for rev\n",
    "                i = ind[-1]\n",
    "                df.loc[i, 'start'] = df.loc[i, 'tes']\n",
    "\n",
    "            # for exon can probably just take [1] or [-1] of temp to access\n",
    "            # first or last exon respectively\n",
    "\n",
    "        # fwd: swap out transcript \"start\" and first exon \"start\" \n",
    "        # rev: swap out transcript \"stop\" and first exon \"stop\"                 \n",
    "#         elif how == 'tss':\n",
    "#             # transcript entry for fwd\n",
    "#             ind = df.loc[(df.strand=='+')&(df.transcript_id==tid)].index.tolist()\n",
    "#             if ind: \n",
    "#                 df.loc[ind, 'start'] = df.loc[ind, 'tss']\n",
    "#                 # start of first exon for fwd\n",
    "#                 ind = ind[1]\n",
    "#                 df.loc[ind].start = df.loc[ind].tss\n",
    "#             # transcript entry for rev\n",
    "#             ind = df.loc[(df.strand=='-')&(df.transcript_id==tid)].index.tolist()\n",
    "#             if ind:\n",
    "#                 df.loc[ind].stop = df.loc[ind].tss\n",
    "#                 # stop of first exon for rev\n",
    "#                 ind = ind[1]\n",
    "#                 df.loc[ind].stop = df.loc[ind].tss\n",
    "\n",
    "    if test:\n",
    "        print()\n",
    "        print('After editing')\n",
    "        print(df[['transcript_id', 'entry_type', 'strand', 'start', 'stop', how]])\n",
    "        \n",
    "    cols=['chr', 'source', 'entry_type', \\\n",
    "          'start', 'stop', 'score', 'strand',\\\n",
    "           'frame', 'fields']\n",
    "    df = df[cols]\n",
    "    fname = '{}_revised_tes.gtf'.format(opref)\n",
    "    df.to_csv(fname, sep='\\t', header=None, index=False, quoting=csv.QUOTE_NONE)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "pleasant-sydney",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "     transcript_id      tes\n",
      "229         142372  4899923\n",
      "304         142634  3344252\n",
      "\n",
      "Before editing\n",
      "        transcript_id  entry_type strand    start     stop        tes\n",
      "92             142372  transcript      +  4857814  4897905  4899923.0\n",
      "93             142372        exon      +  4857814  4857976  4899923.0\n",
      "94             142372        exon      +  4867470  4867532  4899923.0\n",
      "95             142372        exon      +  4878027  4878132  4899923.0\n",
      "96             142372        exon      +  4886744  4886831  4899923.0\n",
      "97             142372        exon      +  4889460  4889602  4899923.0\n",
      "98             142372        exon      +  4890740  4890796  4899923.0\n",
      "99             142372        exon      +  4891915  4892069  4899923.0\n",
      "100            142372        exon      +  4893417  4893563  4899923.0\n",
      "101            142372        exon      +  4894934  4895005  4899923.0\n",
      "102            142372        exon      +  4896356  4897905  4899923.0\n",
      "228081         142634  transcript      -  3344256  3371376  3344252.0\n",
      "228082         142634        exon      -  3370969  3371376  3344252.0\n",
      "228083         142634        exon      -  3360451  3360586  3344252.0\n",
      "228084         142634        exon      -  3358995  3359104  3344252.0\n",
      "228085         142634        exon      -  3358110  3358298  3344252.0\n",
      "228086         142634        exon      -  3355811  3355916  3344252.0\n",
      "228087         142634        exon      -  3355273  3355469  3344252.0\n",
      "228088         142634        exon      -  3352291  3353463  3344252.0\n",
      "228089         142634        exon      -  3350642  3350773  3344252.0\n",
      "228090         142634        exon      -  3350392  3350448  3344252.0\n",
      "228091         142634        exon      -  3348594  3348659  3344252.0\n",
      "228092         142634        exon      -  3347574  3347748  3344252.0\n",
      "228093         142634        exon      -  3347052  3347107  3344252.0\n",
      "228094         142634        exon      -  3346235  3346392  3344252.0\n",
      "228095         142634        exon      -  3344256  3345884  3344252.0\n",
      "fwd ind\n",
      "[92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102]\n",
      "fixing...\n",
      "rev ind\n",
      "[]\n",
      "fwd ind\n",
      "[]\n",
      "rev ind\n",
      "[228081, 228082, 228083, 228084, 228085, 228086, 228087, 228088, 228089, 228090, 228091, 228092, 228093, 228094, 228095]\n",
      "%%fileixing...\n",
      "\n",
      "After editing\n",
      "        transcript_id  entry_type strand      start       stop        tes\n",
      "92             142372  transcript      +  4857814.0  4899923.0  4899923.0\n",
      "93             142372        exon      +  4857814.0  4857976.0  4899923.0\n",
      "94             142372        exon      +  4867470.0  4867532.0  4899923.0\n",
      "95             142372        exon      +  4878027.0  4878132.0  4899923.0\n",
      "96             142372        exon      +  4886744.0  4886831.0  4899923.0\n",
      "97             142372        exon      +  4889460.0  4889602.0  4899923.0\n",
      "98             142372        exon      +  4890740.0  4890796.0  4899923.0\n",
      "99             142372        exon      +  4891915.0  4892069.0  4899923.0\n",
      "100            142372        exon      +  4893417.0  4893563.0  4899923.0\n",
      "101            142372        exon      +  4894934.0  4895005.0  4899923.0\n",
      "102            142372        exon      +  4896356.0  4899923.0  4899923.0\n",
      "228081         142634  transcript      -  3344252.0  3371376.0  3344252.0\n",
      "228082         142634        exon      -  3370969.0  3371376.0  3344252.0\n",
      "228083         142634        exon      -  3360451.0  3360586.0  3344252.0\n",
      "228084         142634        exon      -  3358995.0  3359104.0  3344252.0\n",
      "228085         142634        exon      -  3358110.0  3358298.0  3344252.0\n",
      "228086         142634        exon      -  3355811.0  3355916.0  3344252.0\n",
      "228087         142634        exon      -  3355273.0  3355469.0  3344252.0\n",
      "228088         142634        exon      -  3352291.0  3353463.0  3344252.0\n",
      "228089         142634        exon      -  3350642.0  3350773.0  3344252.0\n",
      "228090         142634        exon      -  3350392.0  3350448.0  3344252.0\n",
      "228091         142634        exon      -  3348594.0  3348659.0  3344252.0\n",
      "228092         142634        exon      -  3347574.0  3347748.0  3344252.0\n",
      "228093         142634        exon      -  3347052.0  3347107.0  3344252.0\n",
      "228094         142634        exon      -  3346235.0  3346392.0  3344252.0\n",
      "228095         142634        exon      -  3344252.0  3345884.0  3344252.0\n"
     ]
    }
   ],
   "source": [
    "# annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "# gtf = 'Brain_talon.gtf'\n",
    "\n",
    "# tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes', test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "boolean-lemon",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'Brain_talon.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "monthly-dryer",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_df = pd.read_csv(annot, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "traditional-brisbane",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>dataset</th>\n",
       "      <th>genome_build</th>\n",
       "      <th>chrom</th>\n",
       "      <th>read_start</th>\n",
       "      <th>read_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>read_length</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM_subtype</th>\n",
       "      <th>fraction_As</th>\n",
       "      <th>custom_label</th>\n",
       "      <th>allelic_label</th>\n",
       "      <th>start_support</th>\n",
       "      <th>end_support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m54284U_191107_221739/45877683/ccs</td>\n",
       "      <td>PacBio_Cortex_Rep1</td>\n",
       "      <td>mm10_ERCC_SIRV</td>\n",
       "      <td>ERCC-00092</td>\n",
       "      <td>1</td>\n",
       "      <td>1105</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1105</td>\n",
       "      <td>55504</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m54284U_191107_221739/29819866/ccs</td>\n",
       "      <td>PacBio_Cortex_Rep1</td>\n",
       "      <td>mm10_ERCC_SIRV</td>\n",
       "      <td>ERCC-00042</td>\n",
       "      <td>1</td>\n",
       "      <td>1007</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1007</td>\n",
       "      <td>55475</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m54284U_191107_221739/110168264/ccs</td>\n",
       "      <td>PacBio_Cortex_Rep1</td>\n",
       "      <td>mm10_ERCC_SIRV</td>\n",
       "      <td>ERCC-00092</td>\n",
       "      <td>1</td>\n",
       "      <td>1104</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1104</td>\n",
       "      <td>55504</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m54284U_191107_221739/14550884/ccs</td>\n",
       "      <td>PacBio_Cortex_Rep1</td>\n",
       "      <td>mm10_ERCC_SIRV</td>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>1</td>\n",
       "      <td>418</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>418</td>\n",
       "      <td>55453</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "      <td>0.30</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m54284U_191107_221739/104335118/ccs</td>\n",
       "      <td>PacBio_Cortex_Rep1</td>\n",
       "      <td>mm10_ERCC_SIRV</td>\n",
       "      <td>ERCC-00042</td>\n",
       "      <td>1</td>\n",
       "      <td>980</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>986</td>\n",
       "      <td>55475</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "      <td>0.35</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             read_name             dataset    genome_build  \\\n",
       "0   m54284U_191107_221739/45877683/ccs  PacBio_Cortex_Rep1  mm10_ERCC_SIRV   \n",
       "1   m54284U_191107_221739/29819866/ccs  PacBio_Cortex_Rep1  mm10_ERCC_SIRV   \n",
       "2  m54284U_191107_221739/110168264/ccs  PacBio_Cortex_Rep1  mm10_ERCC_SIRV   \n",
       "3   m54284U_191107_221739/14550884/ccs  PacBio_Cortex_Rep1  mm10_ERCC_SIRV   \n",
       "4  m54284U_191107_221739/104335118/ccs  PacBio_Cortex_Rep1  mm10_ERCC_SIRV   \n",
       "\n",
       "        chrom  read_start  read_end strand  n_exons  read_length  gene_ID  \\\n",
       "0  ERCC-00092           1      1105      +        1         1105    55504   \n",
       "1  ERCC-00042           1      1007      +        1         1007    55475   \n",
       "2  ERCC-00092           1      1104      +        1         1104    55504   \n",
       "3  ERCC-00004           1       418      +        1          418    55453   \n",
       "4  ERCC-00042           1       980      +        1          986    55475   \n",
       "\n",
       "   ...  annot_gene_name annot_transcript_name gene_novelty transcript_novelty  \\\n",
       "0  ...             None                  None        Known              Known   \n",
       "1  ...             None                  None        Known              Known   \n",
       "2  ...             None                  None        Known              Known   \n",
       "3  ...             None                  None        Known              Known   \n",
       "4  ...             None                  None        Known              Known   \n",
       "\n",
       "  ISM_subtype fraction_As custom_label allelic_label  start_support  \\\n",
       "0        None        0.00         None          None           None   \n",
       "1        None        0.00         None          None           None   \n",
       "2        None        0.00         None          None           None   \n",
       "3        None        0.30         None          None           None   \n",
       "4        None        0.35         None          None           None   \n",
       "\n",
       "  end_support  \n",
       "0        None  \n",
       "1        None  \n",
       "2        None  \n",
       "3        None  \n",
       "4        None  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "electrical-academy",
   "metadata": {},
   "outputs": [],
   "source": [
    "beep = annot_df[['read_name', 'strand', 'transcript_novelty', 'transcript_ID']]\n",
    "beep = beep.groupby(['transcript_ID', 'strand', 'transcript_novelty']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "narrative-spanish",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>read_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>Known</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>Known</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>Known</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   transcript_ID transcript_novelty  read_name\n",
       "0              1              Known          2\n",
       "1              4              Known          3\n",
       "2              7              Known          1\n",
       "3             15              Known          1\n",
       "4             26              Known          2"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "mounted-receptor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>read_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36945</th>\n",
       "      <td>141935</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36947</th>\n",
       "      <td>141937</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36949</th>\n",
       "      <td>141939</td>\n",
       "      <td>+</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36950</th>\n",
       "      <td>141940</td>\n",
       "      <td>+</td>\n",
       "      <td>NIC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36951</th>\n",
       "      <td>141941</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549329</th>\n",
       "      <td>654319</td>\n",
       "      <td>+</td>\n",
       "      <td>NIC</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549332</th>\n",
       "      <td>654322</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549333</th>\n",
       "      <td>654323</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549334</th>\n",
       "      <td>654324</td>\n",
       "      <td>+</td>\n",
       "      <td>ISM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549340</th>\n",
       "      <td>654330</td>\n",
       "      <td>+</td>\n",
       "      <td>ISM</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103152 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_ID strand transcript_novelty  read_name\n",
       "36945          141935      +                NNC          4\n",
       "36947          141937      +                NNC          8\n",
       "36949          141939      +         Intergenic          2\n",
       "36950          141940      +                NIC          3\n",
       "36951          141941      +                NNC         32\n",
       "...               ...    ...                ...        ...\n",
       "549329         654319      +                NIC         27\n",
       "549332         654322      +                NNC          2\n",
       "549333         654323      +                NNC          2\n",
       "549334         654324      +                ISM          3\n",
       "549340         654330      +                ISM         11\n",
       "\n",
       "[103152 rows x 4 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep.loc[(beep.transcript_novelty!='Known')&(beep.read_name > 1)&(beep.strand == '+')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "metropolitan-group",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>read_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36948</th>\n",
       "      <td>141938</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36953</th>\n",
       "      <td>141943</td>\n",
       "      <td>-</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36970</th>\n",
       "      <td>141960</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36979</th>\n",
       "      <td>141969</td>\n",
       "      <td>-</td>\n",
       "      <td>NNC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36982</th>\n",
       "      <td>141972</td>\n",
       "      <td>-</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549312</th>\n",
       "      <td>654302</td>\n",
       "      <td>-</td>\n",
       "      <td>NNC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549314</th>\n",
       "      <td>654304</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549316</th>\n",
       "      <td>654306</td>\n",
       "      <td>-</td>\n",
       "      <td>NNC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549324</th>\n",
       "      <td>654314</td>\n",
       "      <td>-</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549338</th>\n",
       "      <td>654328</td>\n",
       "      <td>-</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99990 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_ID strand transcript_novelty  read_name\n",
       "36948          141938      -                ISM          2\n",
       "36953          141943      -         Intergenic         57\n",
       "36970          141960      -                ISM          4\n",
       "36979          141969      -                NNC          5\n",
       "36982          141972      -            Genomic          5\n",
       "...               ...    ...                ...        ...\n",
       "549312         654302      -                NNC          3\n",
       "549314         654304      -                ISM          3\n",
       "549316         654306      -                NNC          2\n",
       "549324         654314      -            Genomic          6\n",
       "549338         654328      -          Antisense          6\n",
       "\n",
       "[99990 rows x 4 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep.loc[(beep.transcript_novelty!='Known')&(beep.read_name > 1)&(beep.strand == '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "checked-graham",
   "metadata": {},
   "outputs": [],
   "source": [
    "pass_list = pd.read_csv('mouse_brain_whitelist.csv', header=None)\n",
    "pass_list.columns = ['gene_id', 'transcript_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "excessive-commonwealth",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>55504</td>\n",
       "      <td>141821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>55475</td>\n",
       "      <td>141792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>55453</td>\n",
       "      <td>141770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>55463</td>\n",
       "      <td>141780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55495</td>\n",
       "      <td>141812</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id  transcript_id\n",
       "0    55504         141821\n",
       "1    55475         141792\n",
       "2    55453         141770\n",
       "3    55463         141780\n",
       "4    55495         141812"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pass_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "polished-venezuela",
   "metadata": {},
   "outputs": [],
   "source": [
    "beep = beep.loc[beep.transcript_ID.isin(pass_list.transcript_id.tolist())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "comparative-north",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>read_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37382</th>\n",
       "      <td>142372</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37446</th>\n",
       "      <td>142436</td>\n",
       "      <td>+</td>\n",
       "      <td>NIC</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37478</th>\n",
       "      <td>142468</td>\n",
       "      <td>+</td>\n",
       "      <td>NIC</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37479</th>\n",
       "      <td>142469</td>\n",
       "      <td>+</td>\n",
       "      <td>NIC</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37598</th>\n",
       "      <td>142588</td>\n",
       "      <td>+</td>\n",
       "      <td>NIC</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548038</th>\n",
       "      <td>653028</td>\n",
       "      <td>+</td>\n",
       "      <td>ISM</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548042</th>\n",
       "      <td>653032</td>\n",
       "      <td>+</td>\n",
       "      <td>ISM</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548050</th>\n",
       "      <td>653040</td>\n",
       "      <td>+</td>\n",
       "      <td>ISM</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548594</th>\n",
       "      <td>653584</td>\n",
       "      <td>+</td>\n",
       "      <td>NNC</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548920</th>\n",
       "      <td>653910</td>\n",
       "      <td>+</td>\n",
       "      <td>ISM</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3024 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_ID strand transcript_novelty  read_name\n",
       "37382          142372      +                NNC        157\n",
       "37446          142436      +                NIC         74\n",
       "37478          142468      +                NIC         85\n",
       "37479          142469      +                NIC         28\n",
       "37598          142588      +                NIC         94\n",
       "...               ...    ...                ...        ...\n",
       "548038         653028      +                ISM        114\n",
       "548042         653032      +                ISM         47\n",
       "548050         653040      +                ISM        252\n",
       "548594         653584      +                NNC         37\n",
       "548920         653910      +                ISM        121\n",
       "\n",
       "[3024 rows x 4 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep.loc[(beep.transcript_novelty!='Known')&(beep.read_name > 1)&(beep.strand == '+')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "fitting-water",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>read_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37272</th>\n",
       "      <td>142262</td>\n",
       "      <td>-</td>\n",
       "      <td>NNC</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37644</th>\n",
       "      <td>142634</td>\n",
       "      <td>-</td>\n",
       "      <td>NIC</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37645</th>\n",
       "      <td>142635</td>\n",
       "      <td>-</td>\n",
       "      <td>NIC</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37846</th>\n",
       "      <td>142836</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37847</th>\n",
       "      <td>142837</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547976</th>\n",
       "      <td>652966</td>\n",
       "      <td>-</td>\n",
       "      <td>NIC</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547981</th>\n",
       "      <td>652971</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547985</th>\n",
       "      <td>652975</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547986</th>\n",
       "      <td>652976</td>\n",
       "      <td>-</td>\n",
       "      <td>ISM</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548488</th>\n",
       "      <td>653478</td>\n",
       "      <td>-</td>\n",
       "      <td>NIC</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2841 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_ID strand transcript_novelty  read_name\n",
       "37272          142262      -                NNC         38\n",
       "37644          142634      -                NIC         80\n",
       "37645          142635      -                NIC         46\n",
       "37846          142836      -                ISM         30\n",
       "37847          142837      -                ISM         45\n",
       "...               ...    ...                ...        ...\n",
       "547976         652966      -                NIC         37\n",
       "547981         652971      -                ISM         32\n",
       "547985         652975      -                ISM         47\n",
       "547986         652976      -                ISM         21\n",
       "548488         653478      -                NIC        102\n",
       "\n",
       "[2841 rows x 4 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beep.loc[(beep.transcript_novelty!='Known')&(beep.read_name > 1)&(beep.strand == '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "flying-johnson",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
