{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dedicated-canadian",
   "metadata": {},
   "source": [
    "The goal of this analysis is to try and refine the 3' end calling of TALON by using the GENCODE strategy of longest TES per transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "improving-executive",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "serial-amplifier",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_longest_ends(annot, how='tes', novelty='novel'):\n",
    "    df = pd.read_csv(annot, sep='\\t')\n",
    "    \n",
    "    if novelty == 'novel':\n",
    "        df = df.loc[df.transcript_novelty != 'Known']\n",
    "    \n",
    "    fwd = df.loc[df.strand == '+']\n",
    "    rev = df.loc[df.strand == '-']\n",
    "    \n",
    "    # furthest downstream for tes\n",
    "    # if + strand, max coord of read end\n",
    "    # if - strand, min coord of read end\n",
    "    if how == 'tes':\n",
    "        fwd = fwd[['transcript_ID', 'read_end']]\n",
    "        fwd = fwd.groupby('transcript_ID').max().reset_index()\n",
    "        rev = rev[['transcript_ID', 'read_end']]\n",
    "        rev = rev.groupby('transcript_ID').min().reset_index()\n",
    "        \n",
    "        \n",
    "    # furthest upstream for tss:\n",
    "    # if + strand, min coord of read start\n",
    "    # if - strand, max coord of read start\n",
    "    elif how == 'tss':\n",
    "        fwd = fwd[['transcript_ID', 'read_start']]\n",
    "        fwd = fwd.groupby('transcript_ID').min().reset_index()\n",
    "        rev = rev[['transcript_ID', 'read_start']]\n",
    "        rev = rev.groupby('transcript_ID').max().reset_index()\n",
    "    \n",
    "    # concat fwd and rev\n",
    "    df = pd.concat([fwd, rev])\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "closing-grenada",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the longest ends from the read annotation file\n",
    "# annot: TALON read annotation file path\n",
    "# how: 'tss' or 'tes', tss will find start ends and tes will find stop ends\n",
    "# gtf: gtf file location\n",
    "# ends: df with transcript_ID, end coordinate\n",
    "# how: 'tss' or 'tes'\n",
    "# opref: output file prefix\n",
    "def replace_gtf_end_coords(gtf, ends, opref, how='tes', test=False, verbose=False):\n",
    "\n",
    "    # read preexisting GTF and\n",
    "    gtf_df = pd.read_csv(gtf, sep='\\t', header=None, \\\n",
    "                names=['chr', 'source', 'entry_type', \\\n",
    "                       'start', 'stop', 'score', 'strand',\\\n",
    "                       'frame', 'fields'], comment='#')\n",
    "\n",
    "    # get relevant values from fields\n",
    "    gtf_df['transcript_id'] = np.nan\n",
    "    gtf_df.loc[gtf_df.entry_type!='gene', 'transcript_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].fields.str.split(pat='talon_transcript \"', n=1, expand=True)[1]\n",
    "    gtf_df.loc[gtf_df.entry_type!='gene', 'transcript_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].transcript_id.str.split(pat='\"', n=1, expand=True)[0]\n",
    "\n",
    "\n",
    "    if how == 'tes':\n",
    "        ends.columns = ['transcript_id', 'tes']\n",
    "    elif how == 'tss':\n",
    "        ends.columns = ['transcript_id', 'tss']\n",
    "\n",
    "    # merge gtf_df with end information\n",
    "#     ends.transcript_id = ends.transcript_id.astype('str')\n",
    "    df = gtf_df.loc[gtf_df.transcript_id.notnull()]\n",
    "    ends.transcript_id = ends.transcript_id.astype('str')\n",
    "    gtf_df.transcript_id = gtf_df.transcript_id.astype('str')\n",
    "    gtf_df = gtf_df.merge(ends, how='left', on='transcript_id')\n",
    "    df.transcript_id = df.transcript_id.astype('str')\n",
    "    df = df.merge(ends, how='inner')\n",
    "\n",
    "    if test:\n",
    "        print('Before editing')\n",
    "        print(gtf_df[['transcript_id', 'entry_type', 'strand', 'start', 'stop', how]])\n",
    "\n",
    "    # swap out read starts or ends for the longest ones\n",
    "    tids = df.transcript_id.unique()\n",
    "    for t, tid in enumerate(tids):\n",
    "        if t % 1000 == 0 and verbose:\n",
    "            print('Processing transcript {} of {}'.format(t, len(tids)))\n",
    "\n",
    "        # fwd: swap out transcript \"stop\" and last exon \"stop\"\n",
    "        # rev: swap out transcript \"start\" and last exon \"start\"\n",
    "        if how == 'tes':\n",
    "            # tes fwd\n",
    "            ind = gtf_df.loc[(gtf_df.strand=='+')&(gtf_df.transcript_id==tid)].index.tolist()\n",
    "            if ind:\n",
    "                # stop of transcript for fwd\n",
    "                i = ind[0]\n",
    "                gtf_df.loc[i, 'stop'] = gtf_df.loc[i, 'tes']\n",
    "                # stop of last exon for fwd\n",
    "                i = ind[-1]\n",
    "                gtf_df.loc[i, 'stop'] = gtf_df.loc[i, 'tes']\n",
    "\n",
    "            # tes rev\n",
    "            ind = gtf_df.loc[(gtf_df.strand=='-')&(gtf_df.transcript_id==tid)].index.tolist()\n",
    "            if ind:\n",
    "                # start of trancscript for rev\n",
    "                i = ind[0]\n",
    "                gtf_df.loc[i, 'start'] = gtf_df.loc[i, 'tes']\n",
    "                # start of last exon for rev\n",
    "                i = ind[-1]\n",
    "                gtf_df.loc[i, 'start'] = gtf_df.loc[i, 'tes']\n",
    "        # fwd: swap out transcript \"start\" and first exon \"start\"\n",
    "        # rev: swap out transcript \"stop\" and first exon \"stop\"\n",
    "        elif how == 'tss':\n",
    "            # tss fwd\n",
    "            ind = gtf_df.loc[(gtf_df.strand=='+')&(gtf_df.transcript_id==tid)].index.tolist()\n",
    "            if ind:\n",
    "                # start of transcript for fwd\n",
    "                i = ind[0]\n",
    "                gtf_df.loc[i, 'start'] = gtf_df.loc[i, 'tss']\n",
    "                # start of first exon for fwd\n",
    "                i = ind[1]\n",
    "                gtf_df.loc[i, 'start'] = gtf_df.loc[i, 'tss']\n",
    "            # tss rev\n",
    "            ind = gtf_df.loc[(gtf_df.strand=='-')&(gtf_df.transcript_id==tid)].index.tolist()\n",
    "            if ind:\n",
    "                # stop of transcript for rev\n",
    "                i = ind[0]\n",
    "                gtf_df.loc[i, 'stop'] = gtf_df.loc[i, 'tss']\n",
    "                # stop of first exon for rev\n",
    "                i = ind[1]\n",
    "                gtf_df.loc[i, 'stop'] = gtf_df.loc[i, 'tss']\n",
    "    \n",
    "    # now fix gene coordinates\n",
    "    gtf_df['gene_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].fields.str.split(pat='talon_gene \"', n=1, expand=True)[1]\n",
    "    gtf_df['gene_id'] = gtf_df.loc[gtf_df.entry_type!='gene'].gene_id.str.split(pat='\"', n=1, expand=True)[0]\n",
    "\n",
    "    # tes\n",
    "    if how == 'tes':\n",
    "        print('hewwo?')\n",
    "        # fwd: replace \"stop\" of the gene with the maximum of the \"stops\"\n",
    "        gene_ind = gtf_df.loc[(gtf_df.strand == '+')&(gtf_df.entry_type=='gene')&(gtf_df.tes.notnull())].index.tolist()\n",
    "        if gene_ind:\n",
    "            fwd = gtf_df.loc[(gtf_df.strand == '+')&(gtf_df.entry_type=='transcript')]\n",
    "            if test:\n",
    "                print('fwd')\n",
    "                print(gtf_df.loc[gene_ind])\n",
    "            gtf_df.loc[gene_ind, 'stop'] = gtf_df.loc[gene_ind].apply(lambda x: \\\n",
    "                fwd.loc[fwd.gene_id==x.gene_id, 'stop'].max(), axis=1)\n",
    "        \n",
    "        # rev: replace \"start\" of the gene with the minimum of the \"starts\"\n",
    "        gene_ind = gtf_df.loc[(gtf_df.strand == '-')&(gtf_df.entry_type=='gene')&(gtf_df.tes.notnull())].index.tolist()\n",
    "        if gene_ind:\n",
    "            rev = gtf_df.loc[(gtf_df.strand == '-')&(gtf_df.entry_type=='transcript')]\n",
    "            if test: \n",
    "                print('rev')\n",
    "                print(gtf_df.loc[gene_ind])\n",
    "            gtf_df.loc[gene_ind, 'start'] = gtf_df.loc[gene_ind].apply(lambda x: \\\n",
    "                rev.loc[rev.gene_id==x.gene_id, 'start'].min(), axis=1)\n",
    "        \n",
    "    # tss\n",
    "    elif how == 'tss':\n",
    "        # fwd: replace \"start\" of the gene with the minimum of the \"starts\"\n",
    "        gene_ind = gtf_df.loc[(gtf_df.strand == '+')&(gtf_df.entry_type=='gene')&(gtf_df.tss.notnull())].index.tolist()\n",
    "        if gene_ind: \n",
    "            fwd = gtf_df.loc[(gtf_df.strand == '+')&(gtf_df.entry_type=='transcript')]\n",
    "            gtf_df.loc[gene_ind, 'start'] = gtf_df.loc[gene_ind].apply(lambda x: \\\n",
    "                fwd.loc[fwd.gene_id==x.gene_id, 'start'].min(), axis=1)\n",
    "        \n",
    "        # rev: replace \"stop\" of the gene with the maximum of the \"stops\"\n",
    "        gene_ind = gtf_df.loc[(gtf_df.strand == '-')&(gtf_df.entry_type=='gene')&(gtf_df.tss.notnull())].index.tolist()\n",
    "        if gene_ind:\n",
    "            rev = gtf_df.loc[(gtf_df.strand == '-')&(gtf_df.entry_type=='transcript')]\n",
    "            gtf_df.loc[gene_ind, 'stop'] = gtf_df.loc[gene_ind].apply(lambda x: \\\n",
    "                rev.loc[rev.gene_ind==x.gene_ind, 'stop'].max(), axis=1)\n",
    "\n",
    "    if test:\n",
    "        print()\n",
    "        print('After editing')\n",
    "        print(gtf_df[['transcript_id', 'entry_type', 'strand', 'start', 'stop', how]])\n",
    "\n",
    "    cols=['chr', 'source', 'entry_type', \\\n",
    "          'start', 'stop', 'score', 'strand',\\\n",
    "           'frame', 'fields']\n",
    "    gtf_df = gtf_df[cols]\n",
    "    gtf_df['start'] = gtf_df['start'].astype('int')\n",
    "    gtf_df['stop'] = gtf_df['stop'].astype('int')\n",
    "    if test:\n",
    "        fname = '{}_revised_{}_test.gtf'.format(opref, how)\n",
    "    else:\n",
    "        fname = '{}_revised_{}.gtf'.format(opref, how)\n",
    "    gtf_df.to_csv(fname, sep='\\t', header=None, index=False, quoting=csv.QUOTE_NONE)\n",
    "    return gtf_df, fname"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unexpected-cornwall",
   "metadata": {},
   "source": [
    "#### test with 4 transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "minimal-lloyd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before editing\n",
      "   transcript_id  entry_type strand     start      stop         tes\n",
      "0            nan        gene      -   3206523   3215632         NaN\n",
      "1              4  transcript      -   3206523   3215632         NaN\n",
      "2              4        exon      -   3213439   3215632         NaN\n",
      "3              4        exon      -   3206523   3207317         NaN\n",
      "4            nan        gene      +   4857814   4897909         NaN\n",
      "5             63  transcript      +   4857814   4897905         NaN\n",
      "6             63        exon      +   4857814   4857976         NaN\n",
      "7             63        exon      +   4867470   4867532         NaN\n",
      "8             63        exon      +   4878027   4878132         NaN\n",
      "9             63        exon      +   4886744   4886831         NaN\n",
      "10            63        exon      +   4889457   4889602         NaN\n",
      "11            63        exon      +   4890740   4890796         NaN\n",
      "12            63        exon      +   4891915   4892069         NaN\n",
      "13            63        exon      +   4893417   4893563         NaN\n",
      "14            63        exon      +   4894934   4895005         NaN\n",
      "15            63        exon      +   4896356   4897905         NaN\n",
      "16           nan        gene      -  25066282  25228827         NaN\n",
      "17        161766  transcript      -  25066282  25228827  25066282.0\n",
      "18        161766        exon      -  25228421  25228827  25066282.0\n",
      "19        161766        exon      -  25226720  25226823  25066282.0\n",
      "20        161766        exon      -  25221760  25221862  25066282.0\n",
      "21        161766        exon      -  25131251  25131401  25066282.0\n",
      "22        161766        exon      -  25128819  25128888  25066282.0\n",
      "23        161766        exon      -  25126641  25126707  25066282.0\n",
      "24        161766        exon      -  25117289  25117374  25066282.0\n",
      "25        161766        exon      -  25112177  25112275  25066282.0\n",
      "26        161766        exon      -  25111692  25111863  25066282.0\n",
      "27        161766        exon      -  25101433  25101528  25066282.0\n",
      "28        161766        exon      -  25098876  25098915  25066282.0\n",
      "29        161766        exon      -  25093765  25094408  25066282.0\n",
      "30        161766        exon      -  25084244  25084279  25066282.0\n",
      "31        161766        exon      -  25074685  25074789  25066282.0\n",
      "32        161766        exon      -  25066282  25068356  25066282.0\n",
      "hewwo?\n",
      "\n",
      "After editing\n",
      "   transcript_id  entry_type strand       start      stop         tes\n",
      "0            nan        gene      -   3206523.0   3215632         NaN\n",
      "1              4  transcript      -   3206523.0   3215632         NaN\n",
      "2              4        exon      -   3213439.0   3215632         NaN\n",
      "3              4        exon      -   3206523.0   3207317         NaN\n",
      "4            nan        gene      +   4857814.0   4897909         NaN\n",
      "5             63  transcript      +   4857814.0   4897905         NaN\n",
      "6             63        exon      +   4857814.0   4857976         NaN\n",
      "7             63        exon      +   4867470.0   4867532         NaN\n",
      "8             63        exon      +   4878027.0   4878132         NaN\n",
      "9             63        exon      +   4886744.0   4886831         NaN\n",
      "10            63        exon      +   4889457.0   4889602         NaN\n",
      "11            63        exon      +   4890740.0   4890796         NaN\n",
      "12            63        exon      +   4891915.0   4892069         NaN\n",
      "13            63        exon      +   4893417.0   4893563         NaN\n",
      "14            63        exon      +   4894934.0   4895005         NaN\n",
      "15            63        exon      +   4896356.0   4897905         NaN\n",
      "16           nan        gene      -  25066282.0  25228827         NaN\n",
      "17        161766  transcript      -  25066282.0  25228827  25066282.0\n",
      "18        161766        exon      -  25228421.0  25228827  25066282.0\n",
      "19        161766        exon      -  25226720.0  25226823  25066282.0\n",
      "20        161766        exon      -  25221760.0  25221862  25066282.0\n",
      "21        161766        exon      -  25131251.0  25131401  25066282.0\n",
      "22        161766        exon      -  25128819.0  25128888  25066282.0\n",
      "23        161766        exon      -  25126641.0  25126707  25066282.0\n",
      "24        161766        exon      -  25117289.0  25117374  25066282.0\n",
      "25        161766        exon      -  25112177.0  25112275  25066282.0\n",
      "26        161766        exon      -  25111692.0  25111863  25066282.0\n",
      "27        161766        exon      -  25101433.0  25101528  25066282.0\n",
      "28        161766        exon      -  25098876.0  25098915  25066282.0\n",
      "29        161766        exon      -  25093765.0  25094408  25066282.0\n",
      "30        161766        exon      -  25084244.0  25084279  25066282.0\n",
      "31        161766        exon      -  25074685.0  25074789  25066282.0\n",
      "32        161766        exon      -  25066282.0  25068356  25066282.0\n"
     ]
    }
   ],
   "source": [
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'test.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes', test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "dietary-lexington",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before editing\n",
      "   transcript_id  entry_type strand    start     stop        tes\n",
      "0            nan        gene      -  3206523  3215632        NaN\n",
      "1              4  transcript      -  3206523  3215632        NaN\n",
      "2              4        exon      -  3213439  3215632        NaN\n",
      "3              4        exon      -  3206523  3207317        NaN\n",
      "4            nan        gene      +  4857814  4897905        NaN\n",
      "5         142372  transcript      +  4857814  4897905  4899923.0\n",
      "6         142372        exon      +  4857814  4857976  4899923.0\n",
      "7         142372        exon      +  4867470  4867532  4899923.0\n",
      "8         142372        exon      +  4878027  4878132  4899923.0\n",
      "9         142372        exon      +  4886744  4886831  4899923.0\n",
      "10        142372        exon      +  4889460  4889602  4899923.0\n",
      "11        142372        exon      +  4890740  4890796  4899923.0\n",
      "12        142372        exon      +  4891915  4892069  4899923.0\n",
      "13        142372        exon      +  4893417  4893563  4899923.0\n",
      "14        142372        exon      +  4894934  4895005  4899923.0\n",
      "15        142372        exon      +  4896356  4897905  4899923.0\n",
      "16           nan        gene      -  3344256  3409189        NaN\n",
      "17        142634  transcript      -  3344256  3371376  3344252.0\n",
      "18        142634        exon      -  3370969  3371376  3344252.0\n",
      "19        142634        exon      -  3360451  3360586  3344252.0\n",
      "20        142634        exon      -  3358995  3359104  3344252.0\n",
      "21        142634        exon      -  3358110  3358298  3344252.0\n",
      "22        142634        exon      -  3355811  3355916  3344252.0\n",
      "23        142634        exon      -  3355273  3355469  3344252.0\n",
      "24        142634        exon      -  3352291  3353463  3344252.0\n",
      "25        142634        exon      -  3350642  3350773  3344252.0\n",
      "26        142634        exon      -  3350392  3350448  3344252.0\n",
      "27        142634        exon      -  3348594  3348659  3344252.0\n",
      "28        142634        exon      -  3347574  3347748  3344252.0\n",
      "29        142634        exon      -  3347052  3347107  3344252.0\n",
      "30        142634        exon      -  3346235  3346392  3344252.0\n",
      "31        142634        exon      -  3344256  3345884  3344252.0\n",
      "32           nan        gene      -  3206523  3215632        NaN\n",
      "33             4  transcript      -  3206523  3215632        NaN\n",
      "34             4        exon      -  3213439  3215632        NaN\n",
      "35             4        exon      -  3206523  3207317        NaN\n",
      "Processing transcript 0 of 2\n",
      "fwd ind\n",
      "[5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]\n",
      "fixing...\n",
      "rev ind\n",
      "[]\n",
      "fwd ind\n",
      "[]\n",
      "rev ind\n",
      "[17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]\n",
      "fixing...\n",
      "\n",
      "After editing\n",
      "   transcript_id  entry_type strand      start       stop        tes\n",
      "0            nan        gene      -  3206523.0  3215632.0        NaN\n",
      "1              4  transcript      -  3206523.0  3215632.0        NaN\n",
      "2              4        exon      -  3213439.0  3215632.0        NaN\n",
      "3              4        exon      -  3206523.0  3207317.0        NaN\n",
      "4            nan        gene      +  4857814.0  4897905.0        NaN\n",
      "5         142372  transcript      +  4857814.0  4899923.0  4899923.0\n",
      "6         142372        exon      +  4857814.0  4857976.0  4899923.0\n",
      "7         142372        exon      +  4867470.0  4867532.0  4899923.0\n",
      "8         142372        exon      +  4878027.0  4878132.0  4899923.0\n",
      "9         142372        exon      +  4886744.0  4886831.0  4899923.0\n",
      "10        142372        exon      +  4889460.0  4889602.0  4899923.0\n",
      "11        142372        exon      +  4890740.0  4890796.0  4899923.0\n",
      "12        142372        exon      +  4891915.0  4892069.0  4899923.0\n",
      "13        142372        exon      +  4893417.0  4893563.0  4899923.0\n",
      "14        142372        exon      +  4894934.0  4895005.0  4899923.0\n",
      "15        142372        exon      +  4896356.0  4899923.0  4899923.0\n",
      "16           nan        gene      -  3344256.0  3409189.0        NaN\n",
      "17        142634  transcript      -  3344252.0  3371376.0  3344252.0\n",
      "18        142634        exon      -  3370969.0  3371376.0  3344252.0\n",
      "19        142634        exon      -  3360451.0  3360586.0  3344252.0\n",
      "20        142634        exon      -  3358995.0  3359104.0  3344252.0\n",
      "21        142634        exon      -  3358110.0  3358298.0  3344252.0\n",
      "22        142634        exon      -  3355811.0  3355916.0  3344252.0\n",
      "23        142634        exon      -  3355273.0  3355469.0  3344252.0\n",
      "24        142634        exon      -  3352291.0  3353463.0  3344252.0\n",
      "25        142634        exon      -  3350642.0  3350773.0  3344252.0\n",
      "26        142634        exon      -  3350392.0  3350448.0  3344252.0\n",
      "27        142634        exon      -  3348594.0  3348659.0  3344252.0\n",
      "28        142634        exon      -  3347574.0  3347748.0  3344252.0\n",
      "29        142634        exon      -  3347052.0  3347107.0  3344252.0\n",
      "30        142634        exon      -  3346235.0  3346392.0  3344252.0\n",
      "31        142634        exon      -  3344252.0  3345884.0  3344252.0\n",
      "32           nan        gene      -  3206523.0  3215632.0        NaN\n",
      "33             4  transcript      -  3206523.0  3215632.0        NaN\n",
      "34             4        exon      -  3213439.0  3215632.0        NaN\n",
      "35             4        exon      -  3206523.0  3207317.0        NaN\n"
     ]
    }
   ],
   "source": [
    "df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes', test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "seven-repair",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>entry_type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>fields</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3206523.0</td>\n",
       "      <td>3215632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; gene_name \"Xkr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3215632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439.0</td>\n",
       "      <td>3215632</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3206523.0</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857814.0</td>\n",
       "      <td>4897905</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033813.15\"; gene_name \"Tc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source  entry_type      start     stop score strand frame  \\\n",
       "0  chr1  HAVANA        gene  3206523.0  3215632     .      -     .   \n",
       "1  chr1  HAVANA  transcript        NaN  3215632     .      -     .   \n",
       "2  chr1  HAVANA        exon  3213439.0  3215632     .      -     .   \n",
       "3  chr1  HAVANA        exon  3206523.0  3207317     .      -     .   \n",
       "4  chr1  HAVANA        gene  4857814.0  4897905     .      +     .   \n",
       "\n",
       "                                              fields  \n",
       "0  gene_id \"ENSMUSG00000051951.5\"; gene_name \"Xkr...  \n",
       "1  gene_id \"ENSMUSG00000051951.5\"; transcript_id ...  \n",
       "2  gene_id \"ENSMUSG00000051951.5\"; transcript_id ...  \n",
       "3  gene_id \"ENSMUSG00000051951.5\"; transcript_id ...  \n",
       "4  gene_id \"ENSMUSG00000033813.15\"; gene_name \"Tc...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "respected-expression",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "# gtf = 'Brain_talon.gtf'\n",
    "\n",
    "# tes = get_longest_ends(annot, how='tes', novelty='novel')\n",
    "# df = replace_gtf_end_coords(gtf, tes, 'beep', how='tes', test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "legitimate-james",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'get_longest_ends' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-07d82c2ba75e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mgtf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'mouse_brain_cortex_talon.gtf'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mtes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_longest_ends\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannot\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'tes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnovelty\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'novel'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'get_longest_ends' is not defined"
     ]
    }
   ],
   "source": [
    "# cortex\n",
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'mouse_brain_cortex_talon.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "corresponding-venice",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing transcript 0 of 5865\n",
      "Processing transcript 1000 of 5865\n",
      "Processing transcript 2000 of 5865\n",
      "Processing transcript 3000 of 5865\n",
      "Processing transcript 4000 of 5865\n",
      "Processing transcript 5000 of 5865\n"
     ]
    }
   ],
   "source": [
    "cort_df = replace_gtf_end_coords(gtf, tes, 'mouse_brain_cortex', how='tes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "initial-singles",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hippocampus\n",
    "annot = 'PacBio_Brain_talon_read_annot.tsv'\n",
    "gtf = 'mouse_brain_hippocampus_talon.gtf'\n",
    "\n",
    "tes = get_longest_ends(annot, how='tes', novelty='novel')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "urban-peeing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing transcript 0 of 5865\n",
      "Processing transcript 1000 of 5865\n",
      "Processing transcript 2000 of 5865\n",
      "Processing transcript 3000 of 5865\n",
      "Processing transcript 4000 of 5865\n",
      "Processing transcript 5000 of 5865\n"
     ]
    }
   ],
   "source": [
    "hipp_df = replace_gtf_end_coords(gtf, tes, 'mouse_brain_hippocampus', how='tes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "conceptual-makeup",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Cannot convert non-finite values (NA or inf) to integer",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-bcc833101884>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m                        \u001b[0;34m'start'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'stop'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'score'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'strand'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m                        'frame', 'fields'], comment='#')\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mgtf_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'start'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgtf_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'int'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0mgtf_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'stop'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgtf_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'int'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mfname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'cortex_revised_tes.gtf'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m   5541\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5542\u001b[0m             \u001b[0;31m# else, only a single dtype is given\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5543\u001b[0;31m             \u001b[0mnew_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5544\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"astype\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5545\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    593\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"raise\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    594\u001b[0m     ) -> \"BlockManager\":\n\u001b[0;32m--> 595\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"astype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    596\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    597\u001b[0m     def convert(\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, f, align_keys, **kwargs)\u001b[0m\n\u001b[1;32m    404\u001b[0m                 \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    405\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 406\u001b[0;31m                 \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    407\u001b[0m             \u001b[0mresult_blocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_extend_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mapplied\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult_blocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    408\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    592\u001b[0m             \u001b[0mvals1d\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    593\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 594\u001b[0;31m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mastype_nansafe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvals1d\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    595\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mValueError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    596\u001b[0m                 \u001b[0;31m# e.g. astype_nansafe can fail on object-dtype of strings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/dtypes/cast.py\u001b[0m in \u001b[0;36mastype_nansafe\u001b[0;34m(arr, dtype, copy, skipna)\u001b[0m\n\u001b[1;32m    959\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    960\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misfinite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 961\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot convert non-finite values (NA or inf) to integer\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    962\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    963\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot convert non-finite values (NA or inf) to integer"
     ]
    }
   ],
   "source": [
    "# REEEEEEFROAMT\n",
    "gtf = 'mouse_brain_cortex_revised_tes.gtf'\n",
    "gtf_df = pd.read_csv(gtf, sep='\\t', header=None, \\\n",
    "                names=['chr', 'source', 'entry_type', \\\n",
    "                       'start', 'stop', 'score', 'strand',\\\n",
    "                       'frame', 'fields'], comment='#')\n",
    "gtf_df['start'] = gtf_df.start.astype('int')\n",
    "gtf_df['stop'] = gtf_df.stop.astype('int')\n",
    "fname = 'cortex_revised_tes.gtf'\n",
    "gtf_df.to_csv(fname, sep='\\t', header=None, index=False, quoting=csv.QUOTE_NONE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "employed-minimum",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>entry_type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>fields</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3206523.0</td>\n",
       "      <td>3215632.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; gene_name \"Xkr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3215632.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439.0</td>\n",
       "      <td>3215632.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3207317.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.5\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3680155.0</td>\n",
       "      <td>3681788.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102348.1\"; gene_name \"Gm1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source  entry_type      start       stop score strand frame  \\\n",
       "0  chr1  HAVANA        gene  3206523.0  3215632.0     .      -     .   \n",
       "1  chr1  HAVANA  transcript        NaN  3215632.0     .      -     .   \n",
       "2  chr1  HAVANA        exon  3213439.0  3215632.0     .      -     .   \n",
       "3  chr1  HAVANA        exon        NaN  3207317.0     .      -     .   \n",
       "4  chr1  HAVANA        gene  3680155.0  3681788.0     .      +     .   \n",
       "\n",
       "                                              fields  \n",
       "0  gene_id \"ENSMUSG00000051951.5\"; gene_name \"Xkr...  \n",
       "1  gene_id \"ENSMUSG00000051951.5\"; transcript_id ...  \n",
       "2  gene_id \"ENSMUSG00000051951.5\"; transcript_id ...  \n",
       "3  gene_id \"ENSMUSG00000051951.5\"; transcript_id ...  \n",
       "4  gene_id \"ENSMUSG00000102348.1\"; gene_name \"Gm1...  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "institutional-steel",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
